File: app/src/main/java/com/example/lpiem/hearthstonecollectorapp/Manager/APISingleton.java
package com.example.lpiem.hearthstonecollectorapp.Manager;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class APISingleton {

    private static String URL_API = "http://10.0.2.2:8000";
    private static APIInterface hearthstoneInstance;

    public static APIInterface getInstance(){
        if (hearthstoneInstance == null)
            synchronized (APISingleton.class){
                createApiBuilder();
            }
            return hearthstoneInstance;
    }


    private static void createApiBuilder(){
        OkHttpClient.Builder httpClient = new OkHttpClient.Builder();

        Gson gson = new GsonBuilder()
                .create();

        Retrofit.Builder builder = new Retrofit.Builder()
                .baseUrl(URL_API)
                .addConverterFactory(GsonConverterFactory.create(gson));

        Retrofit retrofit = builder
                .client(httpClient.build())
                .build();

        hearthstoneInstance = retrofit.create(APIInterface.class);
    }
}

File: app/src/main/java/com/example/lpiem/hearthstonecollectorapp/Manager/APISingleton.kt
package com.example.lpiem.hearthstonecollectorapp.Manager

import com.example.lpiem.hearthstonecollectorapp.Interface.APIInterface
import com.google.gson.GsonBuilder

import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object APISingleton {
    private val URL_API = "http://172.31.247.232:8000/" //"http://nilsw-iem.ddns.net"
    var hearthstoneInstance: APIInterface? = null
    get() {
        if (field == null)
            createApiBuilder()

        return field
    }


    private fun createApiBuilder() { // dans l'init ?
        val httpClient = OkHttpClient.Builder()

        val gson = GsonBuilder()
                .create()

        val builder = Retrofit.Builder()
                .baseUrl(URL_API)
                .addConverterFactory(GsonConverterFactory.create(gson))

        val retrofit = builder
                .client(httpClient.build())
                .build()

        hearthstoneInstance = retrofit.create(APIInterface::class.java)
    }


}
